File: ben/prolog/P31-P41-arithmetic/p39.pl
% A list of prime numbers.

prime_list(A, B, L) :- 
    numlist(2, B, L0), 
    eratosthenes(L0, L1), 
    include([X] >> (X >= A), L1, L).

eratosthenes([], []).
eratosthenes([X|Xs], [X|R]) :- 
    remove_multiples(X, Xs, Xss), 
    eratosthenes(Xss, R).

remove_multiples(_, [], []).
remove_multiples(X, [Y|Ys], R) :-
    0 is Y mod X,
    remove_multiples(X, Ys, R).
remove_multiples(X, [Y|Ys], [Y|R]) :-
    \+ 0 is Y mod X,
    remove_multiples(X, Ys, R).

File: ben/haskell/Q31-41-arithmetic/p39.hs
{-
A list of prime numbers.

Given a range of integers by its lower and upper limit, construct a list of all prime numbers in that range.
-}

primesR :: Int -> Int -> [Int]
primesR a = filter (>= a) . eratosthenes

eratosthenes :: Int -> [Int]
eratosthenes nb
    | nb < 2 = []
    | otherwise = removeMultiples [2..nb]

removeMultiples :: [Int] -> [Int]
removeMultiples [] = []
removeMultiples (x:xs) = x : (removeMultiples . filter (\a -> mod a x /= 0) $ xs)
